File: undo_tool.py
from pathlib import Path
import shutil
from check_file_permission import check_file_permission

def undo_edit(file_path, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    undo_path = file_path.with_suffix(file_path.suffix + ".undo")
    if not undo_path.exists():
        raise ValueError(f"No undo available for {file_path}")
    shutil.move(undo_path, file_path)
    return f"Restored {file_path} from undo"

File: ts-tools/undo_tool.ts

import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

export function undoEdit(filePath: string, config: any): Promise<string> {
    return new Promise((resolve, reject) => {
        checkFilePermission(filePath, "write", config)
            .then(() => {
                const resolvedPath = path.resolve(filePath);
                const undoPath = resolvedPath + ".undo";

                fs.access(undoPath, fs.constants.F_OK, (err) => {
                    if (err) {
                        reject(new Error(`No undo available for ${filePath}`));
                    } else {
                        fs.rename(undoPath, resolvedPath, (err) => {
                            if (err) {
                                reject(err);
                            } else {
                                resolve(`Restored ${filePath} from undo`);
                            }
                        });
                    }
                });
            })
            .catch(reject);
    });
}
